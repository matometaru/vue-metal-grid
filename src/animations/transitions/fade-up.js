export const appear = (rect) => ({
  translateY: rect.top + 10,
  opacity: 0,
});

export const appeared = () => ({ opacity: 1 });

export const enter = appeared;

export const entered = appeared;

export const leaved = appear;