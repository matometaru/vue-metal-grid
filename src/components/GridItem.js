import Vue from 'vue'
import { transition, buildStyles } from '../utils/style-helper';

const Props = {
	index: Number,
	itemKey: [Number, String],
	component: {
		type: String,
		default: 'span'
	},
	rect: Object,
	containerSize: Object,
	duration: Number,
	easing: String,
	appearDelay: Number,
	appear: Function,
	appeared: Function,
	enter: Function,
	entered: Function,
	leaved: Function,
	units: Object,
	vendorPrefix: Boolean,
	userAgent: String,
	mountedCb: Function,
	unmountCb: Function,
	rtl: Boolean,
}

const getTransitionStyles = (type, props) => {
	const { rect, containerSize, index } = props;

	return props[type](rect, containerSize, index);
};

const getPositionStyles = (rect, zIndex, rtl) => ({
	translateX: `${rtl ? -Math.round(rect.left) : Math.round(rect.left)}px`,
	translateY: `${Math.round(rect.top)}px`,
	zIndex,
});

export default Vue.extend({

	name: "GridItem",

	props: Props,

	data() {
		return {
			appearTimer: null,
			node: null,
			state: {
				...getPositionStyles(this.$props.rect, 1, this.$props.rtl),
				...getTransitionStyles('appear', this.$props),
			}
		}
	},

	watch: {
		rect: function() {
			this.setStateIfNeeded({
				...this.state,
				...getPositionStyles(this.$props.rect, 2, this.$props.rtl),
			});
		}
	},

	mounted() {
		setTimeout(
			this.setAppearedStyles
		, this.$props.appearDelay * this.$props.index );
		this.$props.mountedCb(this);
	},

	destroyed() {
		this.$props.unmountCb(this);
	},

	methods: {

		/**
		 * 状態セット
		 */
		setStateIfNeeded(state) {
			this.state = state;
		},

		/**
		 * 初回表示時の状態設定
		 */
		setAppearedStyles() {
			this.setStateIfNeeded({
				...this.state,
				...getTransitionStyles('appeared', this.$props),
				...getPositionStyles(this.$props.rect, 1, this.$props.rtl),
			});
		},

		/**
		 * 追加時の状態設定
		 */
		setEnterStyles() {
			this.setStateIfNeeded({
				...this.state,
				...getPositionStyles(this.$props.rect, 2, this.$props.rtl),
				...getTransitionStyles('enter', this.$props),
			});
		},

		/**
		 * 追加完了時の状態設定
		 */
		setEnteredStyles() {
			this.setStateIfNeeded({
				...this.state,
				...getTransitionStyles('entered', this.$props),
				...getPositionStyles(this.$props.rect, 1, this.$props.rtl),
			});
		},

		/**
		 * 削除時の状態設定
		 */
		setLeaveStyles() {
			this.setStateIfNeeded({
				...this.state,
				...getPositionStyles(this.$props.rect, 2, this.$props.rtl),
				...getTransitionStyles('leaved', this.$props),
			});
		},

		/**
		 * スタイルを返す
		 *
		 * @return {Object} { zIndex: 1, opacity: 1, display: "block", transform: "translateX(15px) translateY(0px)", … }
		 */
		getStyles() {
			const {
				itemKey,
				rect,
				duration,
				easing,
				units,
				vendorPrefix,
				userAgent,
				rtl,
				...rest
			} = this.$props;

			const styles = buildStyles({
				...this.state,
				display: 'block',
				position: 'absolute',
				top: 0,
      			...(rtl ? { right: 0 } : { left: 0 }),
				width: `${rect.width}px`,
				transition: transition(['opacity', 'transform'], duration, easing),
			}, units, vendorPrefix, userAgent);

			return styles;
		},

		/**
		 * styleオブジェクトに設定する
		 */
		setStyles(el,styles) {
			for(let key in styles) {
				el.style[key] = styles[key]; 
			}
		},

		/**
		 * アニメーションフック
		 * enter, afterEnter, leave
		 */
		onBeforeEnter(el) {
			this.setEnterStyles();
		},

		onEnter(el, done) {
			this.setEnteredStyles();
		},

		onLeave(el, done) {
			this.setLeaveStyles();
			// stylesのオブジェクトを取得
			const styles = this.getStyles();
			// elにスタイル設定。
			this.setStyles(el,styles);
			// setTimeoutでduration後にdoneするか、transitionendイベントで完了。 
			setTimeout(done, this.$props.duration);
		},

	},

	render (h) {

		const {
			index,
			component: Element,
			containerSize,
			appearDelay,
			appear,
			appeared,
			enter,
			entered,
			leaved,
			itemKey,
			rect,
			duration,
			easing,
			units,
			vendorPrefix,
			userAgent,
			rtl,
			...rest
		} = this.$props;

		const style = this.getStyles();

		return (
			<transition-plus
				beforeEnter={this.onBeforeEnter}
				onEnter={this.onEnter}
				onLeave={this.onLeave}>
				<Element
					style={style}
				>
					{this.$slots.default}
				</Element>
			</transition-plus>
		)
	}

})