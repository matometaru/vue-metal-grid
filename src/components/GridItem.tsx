import Vue, { PropType } from 'vue';
import { transition, buildStyles, Units } from '../utils/style-helper';

import { RecordPropsDefinition } from 'vue/types/options';

// https://gist.github.com/wonderful-panda/3156681f25ee72e1a3bfbeaf3764288b propsの型を抽出するやつ
export type RequiredPropNames<PD extends RecordPropsDefinition<any>> = ({
  [K in keyof PD]: PD[K] extends { required: true } ? K : never
})[keyof PD];

export type OptionalPropNames<PD extends RecordPropsDefinition<any>> = {
  [K in keyof PD]: PD[K] extends { required: true } ? never : K
}[keyof PD];

export type OuterProps<
  PropDefs extends RecordPropsDefinition<any>
> = PropDefs extends RecordPropsDefinition<infer P>
  ? { [K in RequiredPropNames<PropDefs> & keyof P]: P[K] } &
      { [K in OptionalPropNames<PropDefs> & keyof P]?: P[K] }
  : never;

type Props = OuterProps<typeof Props>;

const Props = {
  index: {
    type: Number,
    required: true as true,
  },
  itemKey: [Number, String],
  component: {
    type: String,
    default: 'span',
  },
  rect: {
    type: Object as PropType<Rect>,
    required: true as true,
  },
  containerSize: Object as PropType<any>,
  duration: Number,
  easing: String,
  appearDelay: Number,
  appear: {
    type: Function,
    required: true as true,
  },
  appeared: {
    type: Function,
    required: true as true,
  },
  enter: {
    type: Function,
    required: true as true,
  },
  entered: {
    type: Function,
    required: true as true,
  },
  leaved: {
    type: Function,
    required: true as true,
  },
  units: {
    type: Object as PropType<Units>,
    required: true as true,
  },
  vendorPrefix: {
    type: Boolean,
    required: true as true,
  },
  mountedCb: Function,
  unmountCb: Function,
  rtl: Boolean,
};

const getTransitionStyles = (type: TransitionType, props: Props) => {
  const { rect, containerSize, index } = props;

  return props[type](rect, containerSize, index);
};

const getPositionStyles = (rect: Rect, zIndex: number, rtl: boolean) => ({
  translateX: `${rtl ? -Math.round(rect.left) : Math.round(rect.left)}px`,
  translateY: `${Math.round(rect.top)}px`,
  zIndex,
});

export default Vue.extend({
  name: 'GridItem',

  props: Props,

  data() {
    return {
      appearTimer: null,
      node: null,
      state: {
        ...getPositionStyles(this.rect, 1, this.rtl),
        ...getTransitionStyles('appear', this.$props as Props),
      },
    };
  },

  watch: {
    rect() {
      this.setStateIfNeeded({
        ...this.state,
        ...getPositionStyles(this.rect, 2, this.rtl),
      });
    },
  },

  mounted() {
    setTimeout(this.setAppearedStyles, this.appearDelay * this.index);
    this.mountedCb(this);
  },

  destroyed() {
    this.unmountCb(this);
  },

  methods: {
    setStateIfNeeded(state: any) {
      this.state = state;
    },

    setAppearedStyles() {
      this.setStateIfNeeded({
        ...this.state,
        ...getTransitionStyles('appeared', this.$props as Props),
        ...getPositionStyles(this.rect, 1, this.rtl),
      });
    },

    setEnterStyles() {
      this.setStateIfNeeded({
        ...this.state,
        ...getPositionStyles(this.rect, 2, this.rtl),
        ...getTransitionStyles('enter', this.$props as Props),
      });
    },

    setEnteredStyles() {
      this.setStateIfNeeded({
        ...this.state,
        ...getTransitionStyles('entered', this.$props as Props),
        ...getPositionStyles(this.rect, 1, this.rtl),
      });
    },

    setLeaveStyles() {
      this.setStateIfNeeded({
        ...this.state,
        ...getPositionStyles(this.rect, 2, this.rtl),
        ...getTransitionStyles('leaved', this.$props as Props),
      });
    },

    /**
     * スタイルを返す
     *
     * @return {Object} { zIndex: 1, opacity: 1, transform: "translateX(15px) translateY(0px)", … }
     */
    getStyles(): Record<string, string> {
      const styles = buildStyles(
        {
          ...this.state,
          display: 'block',
          position: 'absolute',
          top: 0,
          ...(this.rtl ? { right: 0 } : { left: 0 }),
          width: `${this.rect.width}px`,
          transition: transition(
            ['opacity', 'transform'],
            this.duration,
            this.easing
          ),
        },
        this.units,
        this.vendorPrefix
      );

      return styles;
    },

    setStyles(el: HTMLElement, styles: Record<string, string>) {
      for (const key in styles) {
        el.setAttribute(key, styles[key]);
      }
    },

    onBeforeEnter() {
      this.setEnterStyles();
    },

    onEnter() {
      this.setEnteredStyles();
    },

    onLeave(el: HTMLElement, done: any) {
      this.setLeaveStyles();
      const styles = this.getStyles();
      this.setStyles(el, styles);
      setTimeout(done, this.duration);
    },
  },

  render(h) {
    const Element = this.component;
    const style = this.getStyles();

    return (
      // @ts-ignore
      <transition-plus
        beforeEnter={this.onBeforeEnter}
        onEnter={this.onEnter}
        onLeave={this.onLeave}
      >
        <Element style={style}>{this.$slots.default}</Element>
        {/* 
        // @ts-ignore */}
      </transition-plus>
    );
  },
});
