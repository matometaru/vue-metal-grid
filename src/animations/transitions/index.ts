import * as fade from './fade';
import * as fadeDown from './fade-down';
import * as fadeUp from './fade-up';
import * as scaleDown from './scale-down';
import * as scaleUp from './scale-up';
import * as flip from './flip';
import * as helix from './helix';

export { fade, fadeDown, fadeUp, scaleDown, scaleUp, flip, helix };
