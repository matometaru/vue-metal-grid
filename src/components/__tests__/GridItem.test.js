import { mount } from '@vue/test-utils';
import GridItem from '../GridItem';
import * as transitions from '../../animations/transitions';

const mockPorps = {
  index: 1, // 要素のindex
  itemKey: 1, // 要素のkey番号
  component: 'li', // 構成要素
  rect: {
    column: 1,
    height: 1,
    left: 1,
    top: 1,
  },
  containerSize: {
    actualWidth: 100,
    width: 50,
    height: 50,
  },
  duration: 0,
  easing: 'linear',
  appearDelay: 0,
  appear: transitions.fadeUp.appear,
  appeared: transitions.fadeUp.appeared,
  enter: transitions.fadeUp.enter,
  entered: transitions.fadeUp.entered,
  leaved: transitions.fadeUp.leaved,
  units: { length: 'px', angle: 'deg' },
  vendorPrefix: true,
  userAgent: '', // 未使用
  mountedCb: jest.fn(),
  unmountCb: jest.fn(),
  rtl: false,
};

describe('Snapshot Test', () => {
  test('renders correctly', () => {
    const wrapper = mount(GridItem, {
      stubs: ['transition-plus'],
      propsData: mockPorps,
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});

describe('setAppearedStyles', () => {
  test('should be opacity 1', () => {
    const wrapper = mount(GridItem, {
      stubs: ['transition-plus'],
      propsData: mockPorps,
    });
    wrapper.vm.setAppearedStyles();
    expect(wrapper.vm.state.opacity).toBe(1);
  });
});