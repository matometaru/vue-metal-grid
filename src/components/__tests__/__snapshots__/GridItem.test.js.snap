// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test renders correctly 1`] = `
<transition-plus-stub
  beforeenter="function () { [native code] }"
  useragent=""
>
  <li
    style="z-index: 1; opacity: 0; display: block; position: absolute; top: 0px; left: 0px; transition: opacity 0ms linear,transform 0ms linear; transform: translateX(1px) translateY(11px);"
  />
</transition-plus-stub>
`;
